src/java.base/share/classes/sun/nio/ch/NativeSocketAddress.java

```diff
@@ -287,60 +297,47 @@ private InetAddress address(int family) {
 
     /**
      * Stores the given InetAddress in the sin_addr or sin6_addr/sin6_scope_id
-     * fields. For IPv6 addresses, the sin6_addr will be popluated with an
+     * fields. For IPv6 addresses, the sin6_addr will be populated with an
      * IPv4-mapped IPv6 address when the given InetAddress is an IPv4 address.
      */
     private void putAddress(int family, InetAddress ia) {
         if (family == AF_INET) {
             // IPv4 address
-            putAddress(address + OFFSET_SIN4_ADDR, (Inet4Address) ia);
+            putAddress(memory, (Inet4Address) ia);
         } else {
             int scope_id;
             if (ia instanceof Inet4Address) {
                 // IPv4-mapped IPv6 address
-                UNSAFE.setMemory(address + OFFSET_SIN6_ADDR, 10, (byte) 0);
-                UNSAFE.putByte(address + OFFSET_SIN6_ADDR + 10, (byte) 0xff);
-                UNSAFE.putByte(address + OFFSET_SIN6_ADDR + 11, (byte) 0xff);
-                putAddress(address + OFFSET_SIN6_ADDR + 12, (Inet4Address) ia);
+                byte[] ipv6bytes = new byte[16];
+                ipv6bytes[10] = (byte) 0xff;
+                ipv6bytes[11] = (byte) 0xff;
+                byte[] ipv4address = ia.getAddress();
+                ipv6bytes[12] = ipv4address[0];
```

Comment:
Is there a better way to do this?

Reply:
This code seems very suboptimal: there's an heap array creation `new byte[16]`. Then a bunch of heap memory stores. Then a bulk copy off-heap. So we replaced 4 direct off-heap writes with an heap allocation and a reverse bulk copy. That will probably show up in benchmarks.

Reply:
Thanks for the comment. We've tried to apply the following patch for this code:
```
                 // IPv4-mapped IPv6 address
-                byte[] ipv6bytes = new byte[16];
-                ipv6bytes[10] = (byte) 0xff;
-                ipv6bytes[11] = (byte) 0xff;
+                var sin6addrMs = sockaddr_in6.sin6_addr(memory);
+                sin6addrMs.asSlice(0, 10).fill((byte) 0);
                 byte[] ipv4address = ia.getAddress();
-                ipv6bytes[12] = ipv4address[0];
-                ipv6bytes[13] = ipv4address[1];
-                ipv6bytes[14] = ipv4address[2];
-                ipv6bytes[15] = ipv4address[3];
-                MemorySegment heapAddressBytes = MemorySegment.ofArray(ipv6bytes);
-                sockaddr_in6.sin6_addr(memory).copyFrom(heapAddressBytes);
+                sin6addrMs.asSlice(10, 2).fill((byte) 0xff);
+                MemorySegment.copy(ipv4address, 0, sin6addrMs, JAVA_BYTE, 12, 4);
```

And initial benchmark results look promising.